File: binary-tree-preorder-traversal-144.cc
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

// class Solution {
// public:
//   vector<int> preorderTraversal(TreeNode *root) {
//     vector<int> result;
//     TreeNode *curr = root;
//     stack<TreeNode *> s;
//     while (1) {
//       while (curr) {
//         result.push_back(curr->val);
//         if (curr->right) {
//           s.push(curr->right);
//         }
//         curr = curr->left;
//       }
//       if (s.empty()) {
//         break;
//       }
//       curr = s.top();
//       s.pop();
//     }
//     return result;
//   }
// };

class Solution {
public:
  vector<int> preorderTraversal(TreeNode *root) {
    vector<int> result;
    TreeNode *curr = root;
    stack<TreeNode *> s;
    while (curr || !s.empty()) {
      if (curr) { // 一直往左边走
        result.push_back(curr->val);
        if (curr->right) {
          s.push(curr->right);
        }
        curr = curr->left;
      } else { // 控制遍历方向
        curr = s.top();
        s.pop();
      }
    }
    return result;
  }
};

File: binary-tree-preorder-traversal-144.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func preorderTraversal(root *TreeNode) []int {
	var res []int
	var heap []*TreeNode
	curr := root
	for {
		if curr == nil {
			if len(heap) == 0 {
				break
			}
			curr = heap[len(heap)-1]
			heap = heap[:len(heap)-1]
		}
		res = append(res, curr.Val)
		if curr.Right != nil {
			heap = append(heap, curr.Right)
		}
		curr = curr.Left
	}
	return res
}
